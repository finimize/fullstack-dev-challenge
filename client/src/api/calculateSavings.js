import { apolloClient } from '../apolloClient'
import gql from 'graphql-tag'

export const calculateSavings = variables => {
  const query = gql`
    {
      calculations: calculateSavings(
        initialSavings: 0
        monthlySavings: 0
        yearlyInterest: 0
        interestRecurrency: Monthly
      ) {
        month
        amount
      }
    }
  `
  return apolloClient.query({ query, variables }).then(res => res.data)
}
