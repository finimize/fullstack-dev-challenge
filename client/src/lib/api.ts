const BASE_URL = 'http://localhost:3001'

// TODO: Ideally the types in this function would be more generic to allow for reuse
export const fetchData = async (
    apiUrl: string
): Promise<{ savings: number[]; totalInvested: number; interestEarned: number }> => {
    const url = BASE_URL + apiUrl
    const response = await fetch(url, {
        method: 'GET',
        headers: {
            'content-type': 'application/json;charset=UTF-8',
        },
    })

    type JSONResponse = {
        data?: { savings: number[]; totalInvested: number; interestEarned: number }
        errors?: Array<{ message: string }>
    }
    const { data, errors }: JSONResponse = await response.json()

    if (response.ok && data) {
        return Promise.resolve(data)
    } else {
        const error = new Error(errors?.map((e) => e.message).join('\n') ?? 'unknown')
        return Promise.reject(error)
    }
}
