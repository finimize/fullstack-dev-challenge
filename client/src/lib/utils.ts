import type { ChartTooltipItem } from 'chart.js'
import { SAVINGS_DEFAULTS } from './constants'

export const createQueryString = (queries: { [key: string]: any }): string =>
    Object.keys(queries)
        .map((key) => encodeURIComponent(key) + '=' + encodeURIComponent(queries[key]))
        .join('&')

export const appendYearsQuery = (queryString: string) =>
    `${queryString}&years=${SAVINGS_DEFAULTS.YEARS}`

export const extractYearlyData = (projections: number[]) =>
    projections.filter((item, index) => index % 12 === 0)

export const numberFormatter = new Intl.NumberFormat('gb-GB', {
    style: 'currency',
    currency: 'GBP',
})

export const getTooltipTitle = (tooltipItems: ChartTooltipItem[]) => `Year ${tooltipItems[0].label}`

export const getTooltipLabel = (tooltipItem: ChartTooltipItem) => {
    const value = tooltipItem.value ? parseFloat(tooltipItem.value) : 0
    return numberFormatter.format(value)
}
