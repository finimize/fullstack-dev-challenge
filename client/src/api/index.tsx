import { DataValueInterface } from '../store/store.interface'

export const postCalculation = async (
    initialSavings: number,
    interestRate: number,
    compoundingFrequency: number,
    monthlyContributions: number,
): Promise<DataValueInterface> => {
    const response = await fetch('http://localhost:3001/api', {
        method: 'POST',
        headers: {
            Accept: 'application/json',
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            initialSavings,
            interestRate,
            compoundingFrequency,
            monthlyContributions,
        }),
    })
    type JSONResponse = {
        savingsDetails?: DataValueInterface
        errors?: Array<{ message: string }>
    }
    const { savingsDetails, errors }: JSONResponse = (await response.json()) as JSONResponse

    if (response.ok) {
        return savingsDetails ? { ...savingsDetails } : Promise.reject(new Error('InvalidData'))
    }
    const error = new Error(errors?.map((e) => e.message).join('\n') ?? 'unknown')
    return Promise.reject(error)
}
