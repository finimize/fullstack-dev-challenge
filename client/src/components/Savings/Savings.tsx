import React, { FC, useContext, useEffect } from 'react'
import { Box } from '@chakra-ui/react'
import { AppContext } from '../../store'
import { UPDATE_DATA, PREVIOUS_PAGE } from '../../store/types'
import { DataValueInterface } from '../../store/store.interface'
import { PageControls } from '../PageControls'
import { SavingsControls } from '../SavingsControls'
import { SavingsChart } from '../SavingsChart'
import { postCalculation } from '../../api'

export const Savings: FC = () => {
    const { state, dispatch } = useContext(AppContext)
    const { calculations } = state
    const { calculatorMode } = state
    const isSimpleMode = calculatorMode === 'Simple'

    useEffect(() => {
        const monthlyContributionsParam = isSimpleMode
            ? 0
            : Number(calculations.monthlyContributions.value)
        const compoundFreqParam = isSimpleMode ? 1 : calculations.compoundingFrequency
        postCalculation(
            Number(calculations.initialSavings.value),
            Number(calculations.interestRate),
            compoundFreqParam,
            monthlyContributionsParam,
        )
            .then((data: DataValueInterface) =>
                dispatch({
                    type: UPDATE_DATA,
                    payload: {
                        value: { ...data },
                        error: '',
                        isLoading: false,
                    },
                }),
            )
            .catch((err) =>
                dispatch({
                    type: UPDATE_DATA,
                    payload: {
                        value: null,
                        error: String(err),
                        isLoading: false,
                    },
                }),
            )
    }, [
        calculations.interestRate,
        calculations.initialSavings.value,
        calculations.monthlyContributions.value,
        calculations.compoundingFrequency,
    ])

    const prevPage = () => dispatch({ type: PREVIOUS_PAGE })

    return (
        <Box>
            <SavingsControls />
            <SavingsChart />
            <PageControls prevPage={prevPage} nextPage={() => null} />
        </Box>
    )
}
