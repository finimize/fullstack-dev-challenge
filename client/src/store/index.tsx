import React, { createContext, useReducer, FC } from 'react'
import { initialState } from './initialState'
import { reducer } from './reducer'
import {
    UpdateDetailsTypeInterface,
    ValidateDetailsTypeInterface,
    NextPageTypeInterface,
    StateInterface,
    AppProps,
    PreviousPageTypeInterface,
    ToggleModeInterface,
    UpdateCalculationsTypeInterface,
    UpdateInterestRateInterface,
} from './store.interface'

const AppContext = createContext<{
    state: StateInterface
    dispatch: React.Dispatch<
        | UpdateDetailsTypeInterface
        | UpdateCalculationsTypeInterface
        | ValidateDetailsTypeInterface
        | NextPageTypeInterface
        | ToggleModeInterface
        | PreviousPageTypeInterface
        | UpdateInterestRateInterface
    >
}>({
    state: initialState,
    dispatch: () => null,
})

const AppProvider: FC<AppProps> = ({ children }) => {
    const [state, dispatch] = useReducer(reducer, initialState)
    return <AppContext.Provider value={{ state, dispatch }}>{children}</AppContext.Provider>
}

export { AppContext, AppProvider }
