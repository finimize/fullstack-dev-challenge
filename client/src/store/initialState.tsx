import { StateInterface, CompoundingFrequency } from './store.interface'

export const initialState: StateInterface = {
    details: {
        firstName: {
            value: '',
            error: '',
        },
        lastName: {
            value: '',
            error: '',
        },
        email: {
            value: '',
            error: '',
        },
    },
    data: {
        value: null,
        error: '',
        isLoading: false,
    },
    calculations: {
        initialSavings: {
            value: '10',
            error: '',
        },
        interestRate: 20,
        compoundingFrequency: CompoundingFrequency.annually,
        monthlyContributions: {
            value: '0',
            error: '',
        },
    },
    calculatorMode: 'Simple',
    currentPage: 1,
}
