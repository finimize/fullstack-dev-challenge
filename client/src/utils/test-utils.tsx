import React, { FC, useReducer } from 'react'
import { ChakraProvider, extendTheme } from '@chakra-ui/react'
import { render as rtlRender, RenderResult, RenderOptions } from '@testing-library/react'
import { AppContext, DispatchType } from '../store'
import { initialState } from '../store/initialState'
import { reducer } from '../store/reducer'
import { StateInterface } from '../store/store.interface'
import { theme } from '../theme'

const mainTheme = extendTheme(theme)

interface CustomRenderOptions extends Partial<RenderOptions> {
    store?: StateInterface
    defaultTheme?: typeof theme
}

export interface CustomRenderInterface {
    (
        ui: React.ReactElement<unknown, string | React.JSXElementConstructor<unknown>>,
        options: CustomRenderOptions,
    ): { render: RenderResult; dispatchMock: jest.Mock<unknown, unknown[]> }
}

const customRender: CustomRenderInterface = (
    ui,
    { store = initialState, defaultTheme = mainTheme, ...options },
) => {
    const dispatchMock = jest.fn()
    const AllProviders: FC = ({ children }) => {
        // const [calculatorState, setCalculatorState] = useState(store)
        const [calculatorState, dispatch] = useReducer(reducer, store)
        return (
            <ChakraProvider theme={defaultTheme}>
                <AppContext.Provider
                    value={{
                        state: calculatorState,
                        dispatch: dispatchMock.mockImplementation((value: DispatchType) =>
                            dispatch(value),
                        ),
                    }}
                >
                    {' '}
                    {children}{' '}
                </AppContext.Provider>{' '}
            </ChakraProvider>
        )
    }
    return { render: rtlRender(ui, { wrapper: AllProviders, ...options }), dispatchMock }
}

export { customRender }
