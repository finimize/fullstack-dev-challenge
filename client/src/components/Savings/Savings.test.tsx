import React from 'react'
import { cleanup, waitFor } from '@testing-library/react'
import userEvent, { TargetElement } from '@testing-library/user-event'
import { initialState } from '../../store/initialState'
import { customRender } from '../../utils/test-utils'
import * as postMethods from '../../api'
import { Savings } from './Savings'

afterEach(cleanup)

describe('component/Savings', () => {
    it('renders savings component in simple mode', () => {
        const { render } = customRender(<Savings />, {})
        const { container } = render
        expect(container).toMatchSnapshot()
    })

    it('renders savings component in advanced mode', () => {
        const { render } = customRender(<Savings />, {
            store: { ...initialState, calculatorMode: 'Advanced' },
        })
        const { container } = render
        expect(container).toMatchSnapshot()
    })

    it('postCalculation is called', () => {
        const spy = jest.spyOn(postMethods, 'postCalculation')
        const { render } = customRender(<Savings />, {
            store: { ...initialState, calculatorMode: 'Advanced' },
        })
        const { getByTestId } = render
        const montlyContribution = getByTestId('input-Monthly Contributions')
        expect(montlyContribution).toBeInTheDocument()

        userEvent.type(montlyContribution, '6')
        expect(spy).toHaveBeenNthCalledWith(2, 10, 20, 1, 6)
    })

    it('postCalculation is called on render and dispatch is called to update store', async () => {
        const spy = jest.spyOn(postMethods, 'postCalculation')
        spy.mockReturnValue(
            Promise.resolve({
                yearlySavings: [1, 2],
                finalValue: 5,
                yearlyBreakdown: [
                    {
                        yearFinal: 5,
                        savings: [1, 2],
                    },
                ],
            }),
        )

        const { dispatchMock } = customRender(<Savings />, {
            store: { ...initialState, calculatorMode: 'Advanced' },
        })
        await waitFor(() =>
            expect(dispatchMock).toHaveBeenCalledWith({
                payload: {
                    error: '',
                    isLoading: false,
                    value: {
                        finalValue: 5,
                        yearlyBreakdown: [{ savings: [1, 2], yearFinal: 5 }],
                        yearlySavings: [1, 2],
                    },
                },
                type: 'UPDATE_DATA',
            }),
        )
    })

    it('postCalculation is called on render and dispatch is called to update store', async () => {
        const spy = jest.spyOn(postMethods, 'postCalculation')
        spy.mockReturnValue(Promise.reject(new Error()))

        const { dispatchMock } = customRender(<Savings />, {
            store: { ...initialState, calculatorMode: 'Advanced' },
        })
        await waitFor(() =>
            expect(dispatchMock).toHaveBeenCalledWith({
                payload: { error: 'Error', isLoading: false, value: null },
                type: 'UPDATE_DATA',
            }),
        )
    })

    it('allows to hit previous but not next button', async () => {
        const spy = jest.spyOn(postMethods, 'postCalculation')
        spy.mockReturnValue(
            Promise.resolve({
                yearlySavings: [1, 2],
                finalValue: 5,
                yearlyBreakdown: [
                    {
                        yearFinal: 5,
                        savings: [1, 2],
                    },
                ],
            }),
        )
        const { render, dispatchMock } = customRender(<Savings />, {
            store: { ...initialState, calculatorMode: 'Advanced', currentPage: 3 },
        })
        const { getByTestId, queryByTestId } = render
        let prevButton: unknown
        await waitFor(() => {
            prevButton = getByTestId('previous-button')
        })
        let nextButton
        await waitFor(() => {
            nextButton = queryByTestId('next-button')
        })

        expect(nextButton).not.toBeInTheDocument()
        expect(prevButton).toBeInTheDocument()
        userEvent.click(prevButton as TargetElement)
        expect(dispatchMock).toHaveBeenCalledWith({ type: 'PREVIOUS_PAGE' })
    })
})
