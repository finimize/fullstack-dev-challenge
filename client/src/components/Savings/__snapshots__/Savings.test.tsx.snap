// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component/SavingsChart renders savings component in advanced mode 1`] = `
<div>
   
  <div
    class="css-0"
  >
    <div
      class="css-6z87o"
    >
      <p
        class="chakra-text css-nrvl4t"
      >
        Saving controls
      </p>
      <p
        class="chakra-text css-ulf43m"
      >
        Simply change the values below to see your results
      </p>
      <div
        class="css-174mulg"
      >
        <div
          class="css-lyhr6k"
        >
          <div
            class="css-1d1yckv"
          >
            <label
              class="chakra-form__label css-wuh3rg"
            >
              Initial Deposit
            </label>
            <div
              class="css-ylxpfg"
              value="10"
            >
              <input
                aria-invalid="false"
                aria-valuemax="9007199254740991"
                aria-valuemin="0"
                aria-valuenow="10"
                aria-valuetext="10"
                autocomplete="off"
                autocorrect="off"
                class="css-1ttok2p"
                data-testid="input-Initial Deposit"
                inputmode="decimal"
                pattern="[0-9]*(.[0-9]+)?"
                role="spinbutton"
                type="text"
                value="10"
              />
              <div
                aria-hidden="true"
                class="css-1rscer7"
              >
                <div
                  class="css-9pjy17"
                  role="button"
                  tabindex="-1"
                >
                  <svg
                    class="chakra-icon css-onkibi"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12.8,5.4c-0.377-0.504-1.223-0.504-1.6,0l-9,12c-0.228,0.303-0.264,0.708-0.095,1.047 C2.275,18.786,2.621,19,3,19h18c0.379,0,0.725-0.214,0.895-0.553c0.169-0.339,0.133-0.744-0.095-1.047L12.8,5.4z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
                <div
                  class="css-9pjy17"
                  role="button"
                  tabindex="-1"
                >
                  <svg
                    class="chakra-icon css-onkibi"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M21,5H3C2.621,5,2.275,5.214,2.105,5.553C1.937,5.892,1.973,6.297,2.2,6.6l9,12 c0.188,0.252,0.485,0.4,0.8,0.4s0.611-0.148,0.8-0.4l9-12c0.228-0.303,0.264-0.708,0.095-1.047C21.725,5.214,21.379,5,21,5z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <p
              class="chakra-text css-80ht2y"
            />
          </div>
          <div
            class="css-1d1yckv"
          >
            <label
              class="chakra-form__label css-wuh3rg"
            >
              Monthly Contributions
            </label>
            <div
              class="css-ylxpfg"
              value="0"
            >
              <input
                aria-invalid="false"
                aria-valuemax="9007199254740991"
                aria-valuemin="0"
                aria-valuenow="0"
                aria-valuetext="0"
                autocomplete="off"
                autocorrect="off"
                class="css-1ttok2p"
                data-testid="input-Monthly Contributions"
                inputmode="decimal"
                pattern="[0-9]*(.[0-9]+)?"
                role="spinbutton"
                type="text"
                value="0"
              />
              <div
                aria-hidden="true"
                class="css-1rscer7"
              >
                <div
                  class="css-9pjy17"
                  role="button"
                  tabindex="-1"
                >
                  <svg
                    class="chakra-icon css-onkibi"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12.8,5.4c-0.377-0.504-1.223-0.504-1.6,0l-9,12c-0.228,0.303-0.264,0.708-0.095,1.047 C2.275,18.786,2.621,19,3,19h18c0.379,0,0.725-0.214,0.895-0.553c0.169-0.339,0.133-0.744-0.095-1.047L12.8,5.4z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
                <div
                  aria-disabled="true"
                  class="css-9pjy17"
                  disabled=""
                  role="button"
                  tabindex="-1"
                >
                  <svg
                    class="chakra-icon css-onkibi"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M21,5H3C2.621,5,2.275,5.214,2.105,5.553C1.937,5.892,1.973,6.297,2.2,6.6l9,12 c0.188,0.252,0.485,0.4,0.8,0.4s0.611-0.148,0.8-0.4l9-12c0.228-0.303,0.264-0.708,0.095-1.047C21.725,5.214,21.379,5,21,5z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <p
              class="chakra-text css-80ht2y"
            />
          </div>
          <div
            class="css-uarar0"
          >
            <label
              class="chakra-form__label css-wuh3rg"
            >
              Interest Rate
            </label>
            <div
              class="css-x3oexv"
            >
              <p
                class="chakra-text css-1njcsuf"
              >
                0%
              </p>
              <div
                class="chakra-slider css-1jzw7c4"
                style="position: relative; user-select: none; outline: 0; padding-top: 0px; padding-bottom: 0px;"
                tabindex="-1"
              >
                <div
                  class="chakra-slider__track css-yobz93"
                  id="slider-track-3"
                  style="position: absolute; top: 50%; transform: translateY(-50%); width: 100%;"
                >
                  <div
                    class="css-fsxq0m"
                  />
                  <div
                    class="chakra-slider__filled-track css-lbpcfc"
                    style="position: absolute; top: 50%; transform: translateY(-50%); width: 40%; left: 0px;"
                  />
                </div>
                <div
                  aria-orientation="horizontal"
                  aria-valuemax="50"
                  aria-valuemin="0"
                  aria-valuenow="20"
                  class="chakra-slider__thumb css-t6tax2"
                  data-testid="interest-rate-slider"
                  id="slider-thumb-3"
                  role="slider"
                  style="position: absolute; user-select: none; left: calc(40% - 0px);"
                  tabindex="0"
                />
                <input
                  type="hidden"
                  value="20"
                />
              </div>
              <p
                class="chakra-text css-1njcsuf"
              >
                50%
              </p>
            </div>
            <p
              class="chakra-text css-1as9u30"
            >
              Current interest rate: 
              <b>
                20
                %
              </b>
            </p>
          </div>
          <div
            class="css-uarar0"
            data-testid="cfr-radio-input"
          >
            <label
              class="chakra-form__label css-wuh3rg"
            >
              Interest Frequency
            </label>
            <div
              class="css-171bmva"
              role="radiogroup"
            >
              <label
                class="css-l75mtb"
              >
                <input
                  checked=""
                  data-testid="cfr-annually"
                  name="framework"
                  style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                  type="radio"
                  value="annually"
                />
                <div
                  aria-hidden="true"
                  class="css-15px1hi"
                  data-checked=""
                >
                  annually
                </div>
              </label>
              <label
                class="css-l75mtb"
              >
                <input
                  data-testid="cfr-quarterly"
                  name="framework"
                  style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                  type="radio"
                  value="quarterly"
                />
                <div
                  aria-hidden="true"
                  class="css-15px1hi"
                >
                  quarterly
                </div>
              </label>
              <label
                class="css-l75mtb"
              >
                <input
                  data-testid="cfr-monthly"
                  name="framework"
                  style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                  type="radio"
                  value="monthly"
                />
                <div
                  aria-hidden="true"
                  class="css-15px1hi"
                >
                  monthly
                </div>
              </label>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="css-6z87o"
    >
      <p
        class="chakra-text css-nrvl4t"
      >
        Overview
      </p>
    </div>
    <div
      class="css-j4xgao"
    >
      <button
        class="chakra-button css-1mi4eb0"
        data-testid="next-button"
        type="button"
      >
        Next
      </button>
    </div>
  </div>
   
   
</div>
`;

exports[`component/SavingsChart renders savings component in simple mode 1`] = `
<div>
   
  <div
    class="css-0"
  >
    <div
      class="css-6z87o"
    >
      <p
        class="chakra-text css-nrvl4t"
      >
        Saving controls
      </p>
      <p
        class="chakra-text css-ulf43m"
      >
        Simply change the values below to see your results
      </p>
      <div
        class="css-174mulg"
      >
        <div
          class="css-lyhr6k"
        >
          <div
            class="css-1d1yckv"
          >
            <label
              class="chakra-form__label css-wuh3rg"
            >
              Initial Deposit
            </label>
            <div
              class="css-ylxpfg"
              value="10"
            >
              <input
                aria-invalid="false"
                aria-valuemax="9007199254740991"
                aria-valuemin="0"
                aria-valuenow="10"
                aria-valuetext="10"
                autocomplete="off"
                autocorrect="off"
                class="css-1ttok2p"
                data-testid="input-Initial Deposit"
                inputmode="decimal"
                pattern="[0-9]*(.[0-9]+)?"
                role="spinbutton"
                type="text"
                value="10"
              />
              <div
                aria-hidden="true"
                class="css-1rscer7"
              >
                <div
                  class="css-9pjy17"
                  role="button"
                  tabindex="-1"
                >
                  <svg
                    class="chakra-icon css-onkibi"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12.8,5.4c-0.377-0.504-1.223-0.504-1.6,0l-9,12c-0.228,0.303-0.264,0.708-0.095,1.047 C2.275,18.786,2.621,19,3,19h18c0.379,0,0.725-0.214,0.895-0.553c0.169-0.339,0.133-0.744-0.095-1.047L12.8,5.4z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
                <div
                  class="css-9pjy17"
                  role="button"
                  tabindex="-1"
                >
                  <svg
                    class="chakra-icon css-onkibi"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M21,5H3C2.621,5,2.275,5.214,2.105,5.553C1.937,5.892,1.973,6.297,2.2,6.6l9,12 c0.188,0.252,0.485,0.4,0.8,0.4s0.611-0.148,0.8-0.4l9-12c0.228-0.303,0.264-0.708,0.095-1.047C21.725,5.214,21.379,5,21,5z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <p
              class="chakra-text css-80ht2y"
            />
          </div>
          <div
            class="css-uarar0"
          >
            <label
              class="chakra-form__label css-wuh3rg"
            >
              Interest Rate
            </label>
            <div
              class="css-x3oexv"
            >
              <p
                class="chakra-text css-1njcsuf"
              >
                0%
              </p>
              <div
                class="chakra-slider css-1jzw7c4"
                style="position: relative; user-select: none; outline: 0; padding-top: 0px; padding-bottom: 0px;"
                tabindex="-1"
              >
                <div
                  class="chakra-slider__track css-yobz93"
                  id="slider-track-1"
                  style="position: absolute; top: 50%; transform: translateY(-50%); width: 100%;"
                >
                  <div
                    class="css-fsxq0m"
                  />
                  <div
                    class="chakra-slider__filled-track css-lbpcfc"
                    style="position: absolute; top: 50%; transform: translateY(-50%); width: 40%; left: 0px;"
                  />
                </div>
                <div
                  aria-orientation="horizontal"
                  aria-valuemax="50"
                  aria-valuemin="0"
                  aria-valuenow="20"
                  class="chakra-slider__thumb css-t6tax2"
                  data-testid="interest-rate-slider"
                  id="slider-thumb-1"
                  role="slider"
                  style="position: absolute; user-select: none; left: calc(40% - 0px);"
                  tabindex="0"
                />
                <input
                  type="hidden"
                  value="20"
                />
              </div>
              <p
                class="chakra-text css-1njcsuf"
              >
                50%
              </p>
            </div>
            <p
              class="chakra-text css-1as9u30"
            >
              Current interest rate: 
              <b>
                20
                %
              </b>
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      class="css-6z87o"
    >
      <p
        class="chakra-text css-nrvl4t"
      >
        Overview
      </p>
    </div>
    <div
      class="css-j4xgao"
    >
      <button
        class="chakra-button css-1mi4eb0"
        data-testid="next-button"
        type="button"
      >
        Next
      </button>
    </div>
  </div>
   
   
</div>
`;
